packages/financial-templates-lib/src/price-feed/InsuredBridgePriceFeed.ts

```diff
@@ -0,0 +1,172 @@
+  // associated relay for the price request.
+  public async getHistoricalPrice(time: number | string, ancillaryData: string): Promise<BN> {
+    // First get all relay events emitted at the request timestamp.
+    const matchedRelays = this.relays.filter((relay: Relay) => relay.relayTimestamp === time);
+
+    // Next, Parse ancillary data for relay request and find relay if possible with matching params.
+    const parsedAncillaryData: any = parseAncillaryData(ancillaryData);
+    const matchedRelay = matchedRelays.filter(
+      (relay: Relay) =>
+        relay.depositId === parsedAncillaryData.depositId &&
+        relay.depositTimestamp === parsedAncillaryData.depositTimestamp &&
+        relay.recipient.substr(2).toLowerCase() === parsedAncillaryData.recipient &&
+        relay.sender.substr(2).toLowerCase() === parsedAncillaryData.l2Sender &&
+        relay.l1Token.substr(2).toLowerCase() === parsedAncillaryData.l1Token &&
+        relay.amount === parsedAncillaryData.amount.toString() &&
+        relay.slowRelayFeePct === parsedAncillaryData.slowRelayFeePct.toString() &&
+        relay.instantRelayFeePct === parsedAncillaryData.instantRelayFeePct.toString() &&
+        relay.quoteTimestamp === parsedAncillaryData.quoteTimestamp &&
+        relay.realizedLpFeePct === parsedAncillaryData.realizedLpFeePct.toString() &&
+        relay.slowRelayer.substr(2).toLowerCase() === parsedAncillaryData.slowRelayer &&
+        relay.depositContract.substr(2).toLowerCase() === parsedAncillaryData.depositContract
+    );
+
+    // TODO: Do we need to handle the case where all of these params are matched and matchedRelay.length > 1?
+    if (matchedRelay.length === 0) {
+      this.logger.debug({
+        at: "InsuredBridgePriceFeed",
+        message: "No relay event found for price request time",
+        priceRequestTime: time,
+      });
+      return this.toBNWei(isRelayValid.No);
+    }
+
+    const relay = matchedRelay[0];
+
+    // Validate that relay matches a deposit struct exactly.
+    const matchedDeposits = this.deposits.filter(
+      (deposit: Deposit) =>
+        deposit.depositId === relay.depositId &&
+        deposit.depositHash === relay.depositHash &&
+        deposit.timestamp === relay.depositTimestamp &&
+        deposit.sender === relay.sender &&
+        deposit.recipient === relay.recipient &&
+        deposit.l1Token === relay.l1Token &&
+        deposit.amount === relay.amount &&
+        deposit.slowRelayFeePct === relay.slowRelayFeePct &&
+        deposit.instantRelayFeePct === relay.instantRelayFeePct &&
+        deposit.quoteTimestamp === relay.quoteTimestamp
+    );
+    if (matchedDeposits.length > 1) {
+      this.logger.error({
+        at: "InsuredBridgePriceFeed",
+        message: "TODO: Handle multiple deposits associated with same relay attempt",
+      });
+      throw new Error("TODO: Handle multiple relays for same price request timestamp");
```

Comment:
Do we need to log and throw?
- Won't the throwing result in a logger.error somewhere up the stack?
- I thought throwing was better just because we can catch the errors and emit later so it doesn't interfere with other stuff later in the loop, but maybe I'm misunderstanding how our bots typically handle price feed errors.

You and @chrismaree may know better what the _right_ process is here.

Reply:
yep you're right, i'll change that